File: src/main/java/org/example/blind/MergeIntervals.java
package org.example.blind;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

public class MergeIntervals {
    public int[][] merge(int[][] intervals) {
        List<int []> output = new ArrayList<>();
        Arrays.sort(intervals, Comparator.comparingInt(i -> i[0]));
        int start = intervals[0][0];
        int end = intervals[0][1];

        int currStart, currEnd;
        for (int i=1; i < intervals.length; i++) {
            currStart = intervals[i][0];
            currEnd = intervals[i][1];
            if (currStart <= end) {
                end = Math.max(end, currEnd);
            } else {
                output.add(new int[]{start, end});
                start = currStart;
                end = currEnd;
            }
        }

        output.add(new int[]{start, end});
        return output.toArray(new int[output.size()][]);
    }
}

File: src/main/java/org/example/intervals/MergeIntervals.py
from typing import List


class Solution(object):
    def merge(self, intervals):
        intervals.sort(key=lambda i: i[0])
        res = [intervals[0]]

        for start, end in intervals[1:]:
            lastEnd = res[-1][1]

            if start <= lastEnd:
                res[-1][1] = max(lastEnd, end)
            else:
                res.append([start, end])
        return res

    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()

        res = []
        n = len(intervals)
        i = 0
        while i < n:
            start, end = intervals[i]
            while i + 1 < n and end >= intervals[i + 1][0]:
                end = max(end, intervals[i + 1][1])
                i += 1
            res.append([start, end])
            i += 1
        return res
